File: id/src/main/java/com/simprints/id/services/GuidSelectionService.java
package com.simprints.id.services;

import android.app.IntentService;
import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.simprints.id.Application;
import com.simprints.id.data.DataManager;
import com.simprints.id.data.secure.ApiKeyNotFoundException;
import com.simprints.libsimprints.Constants;

import static com.simprints.libsimprints.Constants.SIMPRINTS_SELECTED_GUID;
import static com.simprints.libsimprints.Constants.SIMPRINTS_SESSION_ID;


public class GuidSelectionService extends IntentService {

    private DataManager dataManager;

    public GuidSelectionService() {
        super("GuidSelectionService");
    }

    @Override
    public void onCreate() {
        super.onCreate();
        Application app = ((Application) getApplication());
        dataManager = app.getDataManager();
    }

    private boolean hasValidApiKeyExtra(@NonNull Intent intent) {
        String apiKey = intent.getStringExtra(Constants.SIMPRINTS_API_KEY);

        String dmApiKey;
        try {
            dmApiKey = dataManager.getApiKey();
        } catch (ApiKeyNotFoundException ex) {
            dataManager.logException(ex);
            return false;
        }

        return apiKey != null && apiKey.equals(dmApiKey);
    }

    private boolean hasValidSessionIdExtra(@NonNull Intent intent) {
        String sessionId = intent.getStringExtra(SIMPRINTS_SESSION_ID);
        return sessionId != null && sessionId.equals(dataManager.getSessionId());
    }

    private String getSelectedGuidExtra(@NonNull Intent intent) {
        String selectedGuid = intent.getStringExtra(SIMPRINTS_SELECTED_GUID);
        return selectedGuid != null ? selectedGuid : "null";
    }

    private void onHandleNonNullIntent(@NonNull Intent intent) {
        boolean isValidSelection = hasValidApiKeyExtra(intent) && hasValidSessionIdExtra(intent);
        String selectedGuid = getSelectedGuidExtra(intent);
        if (isValidSelection) {
            dataManager.updateIdentification(selectedGuid);
        }
        dataManager.logGuidSelectionService(selectedGuid, isValidSelection);
    }

    @Override
    protected void onHandleIntent(@Nullable Intent intent) {
        if (intent != null) {
            onHandleNonNullIntent(intent);
        }
    }
}

File: id/src/main/java/com/simprints/id/services/GuidSelectionService.kt
package com.simprints.id.services

import android.annotation.SuppressLint
import android.app.IntentService
import android.content.Intent
import com.simprints.id.Application
import com.simprints.id.data.analytics.crashreport.CrashReportManager
import com.simprints.id.tools.extensions.parseAppConfirmation
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.rxkotlin.subscribeBy
import io.reactivex.schedulers.Schedulers
import timber.log.Timber
import javax.inject.Inject

class GuidSelectionService : IntentService("GuidSelectionService") {

    @Inject lateinit var guidSelectionManager: GuidSelectionManager
    @Inject lateinit var crashReportManager: CrashReportManager

    override fun onCreate() {
        super.onCreate()
        (application as Application).component.inject(this)
    }

    @SuppressLint("CheckResult")
    override fun onHandleIntent(intent: Intent?) {
        if (intent != null) {
            val request = intent.parseAppConfirmation()
            guidSelectionManager.handleIdentityConfirmationRequest(request)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeBy(onComplete = {
                    Timber.d("Added Guid Selection Event")
                })
        }
    }
}
